pylint/checkers/none_ascii_names.py

```diff
@@ -0,0 +1,241 @@
+    Py37Str = str
+else:
+
+    class Py37Str(str):
+        # Allow Python 3.6 compatibility
+        def isascii(self: str) -> bool:
+            return all("\u0000" <= x <= "\u007F" for x in self)
+
+
+class _AsciiOnlyCheckedNode(Protocol):
+    _is_ascii_only: bool
+
+
+class NoneAsciiNamesChecker(BaseChecker):
+    """A strict name checker only allowing ASCII
+
+    If your programming guideline defines that you are programming in English,
+    then there should be no need for non ASCII characters in Python Names.
+    Everybody else can simply not use this check.
+
+    Note: This check only checks Names, so it ignores the content of
+          docstrings and comments!
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+    priority = -1
+
+    msgs = {
+        "C0144": (
+            '%s name "%s" contains a non-ASCII character',
+            "non-ascii-name",
+            (
+                "Used when the name contains at least one non-ASCII unicode character. "
+                "See https://www.python.org/dev/peps/pep-0672/#confusable-characters-in-identifiers"
+                " for a background why this could be bad."
+            ),
+        ),
+    }
+
+    name = "NonASCII-Checker"
+
+    def __init__(self, linter):
+        super().__init__(linter)
+        self._non_ascii_rgx_compiled = re.compile("[^A-Za-z0-9_]")
+
+    def _raise_name_warning(
+        self,
+        node: nodes.NodeNG,
+        node_type: str,
+        name: str,
+        confidence: interfaces.Confidence,
+    ) -> None:
+        type_label = HUMAN_READABLE_TYPES[node_type]
+        args = (type_label.capitalize(), name)
+
+        self.add_message("non-ascii-name", node=node, args=args, confidence=confidence)
+        self.linter.stats.increase_bad_name(node_type, 1)
+
+    def _check_name(
+        self,
+        node_type: str,
+        names: Union[str, List[str]],
+        node: Union[nodes.NodeNG, _AsciiOnlyCheckedNode],
+    ) -> None:
+        """check for a name if it's using non ASCII and set message if it does
+
+        Also set the dynamic attribute ``_is_ascii_only`` that is used to
+        determine if a node has been already checked, so we don't have to handle
+        too many edge cases.
+        """
+        if not isinstance(names, list):
+            names = [names]
+
+        current_state = getattr(node, "_is_ascii_only", True)
+        for name in names:
+            if name is None:
+                # For some nodes i.e. *kwargs from a dict, the name will be empty
+                continue
+            non_ascii_match = self._non_ascii_rgx_compiled.match(name)
+            if not (Py37Str(name).isascii() and non_ascii_match is None):
+                # Note that we require the `.isascii` method as it is fast and
+                # handles the complexities of unicode, so we can use simple regex.
+                self._raise_name_warning(
+                    node,
+                    node_type,
+                    name,
```

Comment:
```suggestion
                    name
```
(Better formatting with black)